Merge gridded_code into master

chaukap opened the pull request:
I am opening this pull request to allow us to review the code as we test it. There are a ton of changes on this branch so it will take us some time to work through everything. 

The biggest lingering issue is Spinup. Therefore, it would be best to start reviewing all files except ST_initialization.c. Once we have a concrete plan for Spinup we can finish the review.

Things to Note
--
* Seven files have been created on this branch.
  * ST_initialization .c and .h deal with initialization code previously in ST_grid.c. They contain all spinup code. There is also a half-finished seed dispersal function. This function should eventually call functions declared in ST_seedDispersal.h, but completion of seed dispersal has been pushed to it's [own milestone](https://github.com/DrylandEcology/STEPWAT2/milestone/26).
  * ST_seedDispersal .c and .h deal with seed dispersal. They are incomplete, and for now only define placeholder functions that print out "Feature coming soon" messages. Eventually the functionality for simulation and initialization seed dispersal will be written here.
  * ST_progressBar .c and .h deal with the new progress bar that I implemented. The new .h file makes it easy to add progress bars to any simulation loop. I therefore ported the progress bar to non-gridded mode, so the whole program can now use the `-p` flag. These files are be 100% complete and bug free (in theory).
  * ST_grid.h was added to provide a list of standard functions that are made to be exported. This way, future programmers to not have to redefine functions from ST_grid.c to use them. They only have to include the .h file.
* All of these files took on functionality previously in ST_grid.c.

chaukap commented on ST_stats.h:
```diff
@@ -0,0 +1,49 @@
+  char *name; /* array of ptrs to names in RGroup & Species */
+  struct accumulators_grid_cell_st *s;    /* array of holding all the years values */
+} typedef GridStatsType;
+
```
We should add the declarations of all exported functions from ST_stats.c. That we we no longer have to re-declare them.

chaukap commented on ST_main.c:
```diff
@@ -164,12 +169,16 @@ int main(int argc, char **argv) {
 	printf("STEPWAT  init_args() executed successfully \n");
 
-        /*
 	if (UseGrid == TRUE) {
```
`if(UseGrid == TRUE)` is redundant. This should be changed to `if(UseGrid)`.

chaukap commented on ST_main.c:
```diff
@@ -187,51 +196,48 @@ int main(int argc, char **argv) {
 	}
 
-	incr = (IntS) ((float) Globals.runModelIterations / 10);
+	incr = (IntS) ((float) SuperGlobals.runModelIterations / 10);
```
`incr` has been deprecated by the new progress bar. It can be removed entirely.

fdpierson commented on ST_environs.c:
```diff
@@ -191,73 +234,73 @@ static void _make_disturbance( void) {
          if (event == NoDisturb) break;
-         Plot.pat_removed = (RandUni(&environs_rng) <= Globals.pat.removal)
+         Plot->pat_removed = (RandUni(&environs_rng) <= Globals->pat.removal)
                            ? TRUE : FALSE;
```
Why use a ternary here when you can just use the boolean expression itself?

chaukap replied:
I've seen quite a few things like this in the code. Good catch!

chaukap replied:
Resolved by 9b99edd.

fdpierson commented on ST_environs.c:
```diff
@@ -191,73 +234,73 @@ static void _make_disturbance( void) {
       case LastDisturb:
       default:
-           Plot.disturbed = (Plot.disturbed) ? Plot.disturbed -1 : 0;
+           Plot->disturbed = (Plot->disturbed) ? Plot->disturbed -1 : 0;
```
You don't need a break here since this is the last case, but it is a good idea to have one here anyways.

chaukap replied:
Resolved by 9b99edd. I also changed `Plot->disturbed -1` to `(Plot->disturbed - 1)` for clarity.

fdpierson commented on ST_mortality.c:
```diff
@@ -345,33 +345,33 @@ void grazing_EndOfYear( void){
 		g = RGroup[rg];
 
-		//printf("inside grazing_EndOfYear() year=%d, rgroupName=%s, grazingfreq_startyr=%d, grazingfreq=%d, proportionGrazing=%f startYear=%d \n",Globals.currYear,g->name,g->grazingfreq_startyr,g->grazingfrq,g->proportion_grazing, RGroup[rg]->startyr);
+		//printf("inside grazing_EndOfYear() year=%d, rgroupName=%s, grazingfreq_startyr=%d, grazingfreq=%d, proportionGrazing=%f startYear=%d \n",Globals->currYear,g->name,g->grazingfreq_startyr,g->grazingfrq,g->proportion_grazing, RGroup[rg]->startyr);
```
Can we get rid of this commented-out code?

kpalmqui replied:
fine with me.

fdpierson commented on ST_mortality.c:
```diff
@@ -381,7 +381,7 @@ void grazing_EndOfYear( void){
 					continue;
 				}
-				//printf( "year=%d calling Species_Proportion_Grazing()  rgroupName=%s, est_count =%d,grazingfreq_startyr=%d, grazingfreq=%d, proportionGrazing=%f \n",Globals.currYear,g->name,RGroup[rg]->est_count,g->grazingfreq_startyr,g->grazingfrq,g->proportion_grazing);
+				//printf( "year=%d calling Species_Proportion_Grazing()  rgroupName=%s, est_count =%d,grazingfreq_startyr=%d, grazingfreq=%d, proportionGrazing=%f \n",Globals->currYear,g->name,RGroup[rg]->est_count,g->grazingfreq_startyr,g->grazingfrq,g->proportion_grazing);
```
Can we get rid of this commented-out code?

kpalmqui replied:
fine with me.

fdpierson commented on ST_params.c:
```diff
@@ -1107,7 +1090,7 @@ static void _species_init( void) {
+      Species[sp]->pseed = pseed / Globals->plotsize;
 /*      Species[sp]->ann_mort_prob = (age > 0)
                                          ? -log(cohort)/age
                                          : 0.0;
```
Can we get rid of this commented-out code?

chaukap replied:
@kpalmqui Is this line something we can remove?

kpalmqui replied:
@chaukap @fdpierson I would liked to retain this for now despite the fact that it is not being used. This appears to be related to establishment of annuals, which we need to revisit. I will open a new issue for this.

kpalmqui replied:
https://github.com/DrylandEcology/STEPWAT2/issues/381

fdpierson commented on ST_resgroups.c:
```diff
@@ -650,7 +651,7 @@ void rgroup_Establish(void) {
                 /* Establishment for species that belong to annual functional groups*/
                 if (Species[sp]->max_age == 1) {
-                    //printf("Globals.currYear = %hu, call to _add_annuals sp=%d Species[sp]->lastyear_relsize : %.5f \n", Globals.currYear, sp, Species[sp]->lastyear_relsize);
+                    //printf("Globals->currYear = %hu, call to _add_annuals sp=%d Species[sp]->lastyear_relsize : %.5f \n", Globals->currYear, sp, Species[sp]->lastyear_relsize);
```
Can we get rid of this commented-out code?

chaukap replied:
@kpalmqui I'm fine with removing this code so the code is more readable. But if you want to keep it for testing on your end that's fine too!

kpalmqui replied:
fine with me.

fdpierson commented on ST_resgroups.c:
```diff
@@ -650,7 +651,7 @@ void rgroup_Establish(void) {
-                    //printf("Globals.currYear = %hu, call to _add_annuals sp=%d Species[sp]->lastyear_relsize : %.5f \n", Globals.currYear, sp, Species[sp]->lastyear_relsize);
+                    //printf("Globals->currYear = %hu, call to _add_annuals sp=%d Species[sp]->lastyear_relsize : %.5f \n", Globals->currYear, sp, Species[sp]->lastyear_relsize);
                     num_est = _add_annuals(rg, sp, Species[sp]->lastyear_relsize);
                     //  printf("g->seedbank annuals=%d \n",g->seedbank);
```
Can we get rid of this commented-out code?

kpalmqui replied:
fine with me.

fdpierson commented on ST_resgroups.c:
```diff
@@ -662,7 +663,7 @@ void rgroup_Establish(void) {
                 if (num_est) {
                     /* printf("%d %d %d %d\n",
-                     Globals.currIter, Globals.currYear, sp, num_est); */
+                     Globals->currIter, Globals->currYear, sp, num_est); */
```
Can we get rid of this commented-out code?

kpalmqui replied:
yes

fdpierson commented on ST_species.c:
```diff
@@ -429,7 +546,7 @@ void save_annual_species_relsize() {
             Species[sp]->lastyear_relsize = getSpeciesRelsize(sp);
             //Species[sp]->lastyear_relsize = 2;
-            //printf("Globals.currYear = %d, sp=%d new updated value lastyear_relsize : %.5f \n", Globals.currYear, sp, Species[sp]->lastyear_relsize);
+            //printf("Globals->currYear = %d, sp=%d new updated value lastyear_relsize : %.5f \n", Globals->currYear, sp, Species[sp]->lastyear_relsize);
```
Can we get rid of this commented-out code?

kpalmqui replied:
yes

fdpierson commented on makefile:
```diff
@@ -98,7 +101,10 @@ EXOBJS	=\
+	$(oDir)/ST_initialization.o\
+	$(oDir)/ST_progressBar.o\
+	$(oDir)/ST_seedDispersal.o
 	#$(oDir)/sxw_tester.o
```
Can we get rid of this commented-out object file? If we need it again, we can add it back later.

chaukap replied:
Resolved by 9b99edd.

fdpierson commented on sxw_soilwat.c:
```diff
@@ -143,11 +137,11 @@ static void _update_transp_coeff(void) {
               l, RGroup[g]->name, g, RGroup[g]->veg_prod_type, getNTranspLayers(RGroup[g]->veg_prod_type),
-              _roots_max[Ilg(l, g)], RGroup[g]->relsize, y->transp_coeff[SW_SHRUB]);
+              SXWResources->_roots_max[Ilg(l, g)], RGroup[g]->relsize, y->transp_coeff[SW_SHRUB]);
              */
```
Can we get rid of this commented-out code?

kpalmqui replied:
I would like to keep this (after we updated the references to relsize)

chaukap commented:
@fdpierson Sorry to get back to you so late, but I am on board with running the code through a formatter. Did you have any specific program in mind?

kpalmqui commented:
@fdpierson @chaukap I also like Freddy's idea about a formatter. Would love to hear what your suggestions are.

fdpierson commented:
@chaukap @kpalmqui I'll get back to you on a program either later today or tomorrow. Since we all probably use different editors, we need a standalone program that can format for us.

fdpierson commented:
@chaukap @kpalmqui After a little bit of searching, it seems like clang-format is quite popular.

kpalmqui commented:
@chaukap @fdpierson thanks Freddy! Chandler - what do you think of that proposal?

also tagging @dschlaep to see what he does in SOILWAT2

dschlaep commented:
@kpalmqui We haven't been using a formatter for SOILWAT2; we have defined a code style (https://github.com/DrylandEcology/DrylandEcologyProtocols#cstyle) but it has not been possible to enforce those with the previous programmers.

I don't know `clang-format`, but it seems an interesting tool. If you decide to use it for STEPWAT2, then I would be interesting to learn about the experience to consider applying it to SOILWAT2 as well.

My experience is that if we cannot add it to CI checks (where a failure would result in github blocking a pull-request), then it will be difficult to maintain the formatting in the future with a different set of programmers who may be less aware and interested.

chaukap commented:
@kpalmqui @fdpierson I like the idea of using a formatter because it is easily maintainable. I just got clang-format working on Visual Studio Code and it fixes all of our indentation issues. 

The only problem I ran into was that the default clang settings try to reorder `#include` statements in alphabetical order. When this happens the compiler fails. That being said I'm sure there's a way to change the settings.

@dschlaep I agree that this would not be 100% enforceable. However, if we use clang-format just once on every file in STEPWAT2 it would have a lasting impact because most functions will not be touched again for months if not years.

kpalmqui commented:
@chaukap @fdpierson I agree it seems like a good idea. Perhaps the next step is to see how we can overcome the problem Chandler outlines above.

chaukap commented:
@fdpierson If you could look into modifying the clang-format settings to be compatible with our code base we could format the files as soon as possible.

@kpalmqui I think we can agree that this formatter is a good idea, but I think we should perform the entire formatting operation on a new branch. Once we run the formatter there will be a ton of indentation changes which would make reviewing the pull request a huge pain.

So I'll open a new issue referencing this pull request where we can continue this discussion.

fdpierson commented:
@chaukap I'll do so at my earliest convenience, although I can't do so today because of work. Any deadlines on this task?

chaukap commented:
@fdpierson No worries! There's no rush on this, and I think we should get gridded_code merged into master before our initial format. So once you have some time take a look at it, and you can post your proposed solution in issue #387

kpalmqui commented:
@fdpierson as we discussed during our last meeting - the top priority task you should be focusing on is finishing your review of gridded code. thanks!

fdpierson commented on ST_mortality.c:
```diff
@@ -348,33 +348,31 @@ void grazing_EndOfYear( void){
-		if (Globals.currYear == grazingyr)
+		if (Globals->currYear == grazingyr)
 		{
 			//printf( "currYear is equal to grazingYear so will iterate all the Species for doing grazing, RGroup[g]->est_count =%d \n",RGroup[rg]->est_count);
```
This line can probably safely be removed.

chaukap replied:
Resolved by 2a1ff37.

fdpierson commented on sxw.c:
```diff
@@ -918,24 +948,24 @@ void _print_debuginfo(void) {
-		//ForEachTranspLayer(t) sum += SXW.transp[Ilp(t,p)];
-		for (t = 0; t < SXW.NSoLyrs; t++)
-			sum += SXW.transpTotal[Ilp(t, p)];
+		//ForEachTranspLayer(t) sum += SXW->transp[Ilp(t,p)];
```
This line can probably safely be removed.

chaukap replied:
Resolved by 2a1ff37.

fdpierson commented on sxw.c:
```diff
@@ -1019,23 +1049,23 @@ void _print_debuginfo(void) {
 			//ForEachTranspLayer(t) {
```
This line can probably safely be removed.

chaukap replied:
Resolved by 103ca00.

fdpierson commented on sxw.c:
```diff
@@ -1019,23 +1049,23 @@ void _print_debuginfo(void) {
 	}
 
 	fprintf(f, "\n------ Transpiration Total Values -------\nPeriod:");
 	//ForEachTranspLayer(t)
```
This line can probably safely be removed.

chaukap replied:
Resolved by 103ca00.

fdpierson commented on sxw.c:
```diff
@@ -1019,23 +1049,23 @@ void _print_debuginfo(void) {
 	ForEachTrPeriod(p)
 	{
 		fprintf(f, "%d : ", p + 1);
 		//ForEachTranspLayer(t)
```
This line can probably safely be removed.

chaukap replied:
Resolved by 103ca00.

fdpierson commented on sxw_resource.c:
```diff
@@ -150,14 +119,14 @@ void _sxw_update_resource(void) {
          * (cm) to biomass that can be supported by those resources (g/cm) */
 	ForEachGroup(g)
 	{
 //printf("for groupName= %smresource_cur prior to multiplication: %f\n",RGroup[g]->name, _resource_cur[g]);
```
This line can probably safely be removed.

chaukap replied:
@fdpierson I case anyone wants to use these print statements I combined the "prior to multiplication" and "post multiplication" statements into a single statement.

fdpierson commented on sxw_resource.c:
```diff
@@ -150,14 +119,14 @@ void _sxw_update_resource(void) {
 //printf("for groupName= %smresource_cur prior to multiplication: %f\n",RGroup[g]->name, _resource_cur[g]);
-		_resource_cur[g] = _resource_cur[g] * _bvt;
+		SXWResources->_resource_cur[g] = SXWResources->_resource_cur[g] * SXWResources->_bvt;
 //printf("for groupName= %s, resource_cur post multiplication: %f\n\n",Rgroup[g]->name, _resource_cur[g]);
```
This line can probably safely be removed.

chaukap replied:
Resolved by 2a1ff37.

fdpierson commented on sxw_resource.c:
```diff
@@ -150,14 +119,14 @@ void _sxw_update_resource(void) {
+		SXWResources->_resource_cur[g] = SXWResources->_resource_cur[g] * SXWResources->_bvt;
 //printf("for groupName= %s, resource_cur post multiplication: %f\n\n",Rgroup[g]->name, _resource_cur[g]);
 	}
 /* _print_debuginfo(); */
```
This line can probably safely be removed.

chaukap replied:
Resolved by 103ca00.

fdpierson commented on sxw_resource.c:
```diff
@@ -270,7 +233,7 @@ static void _transp_contribution_by_group(RealF use_by_group[]) {
+                use_by_group[g] += (RealF) (SXWResources->_roots_active_rel[Iglp(g, l, p)] * transp[Ilp(l, p)]);
             }
         }
         //printf("for groupName= %s, use_by_group[g] in transp= %f \n",RGroup[g]->name,use_by_group[g] );
```
This line can probably safely be removed.

chaukap replied:
Resolved by 2a1ff37.

fdpierson commented on sxw_resource.c:
```diff
@@ -300,101 +263,101 @@ static void _transp_contribution_by_group(RealF use_by_group[]) {
-            RealF min = transp_window.ratio_average - transp_ratio_sd;
-            RealF max = transp_window.ratio_average + transp_ratio_sd;
+    if (transp_ratio < (transp_window->ratio_average - 1 * transp_ratio_sd)) {
+        //printf("Year %d below 1 sd: ratio = %f, average = %f, sd = %f\n", Globals->currYear,transp_ratio,
```
This line can probably safely be removed.

chaukap replied:
Resolved by 103ca00.

fdpierson commented on sxw_resource.c:
```diff
@@ -300,101 +263,101 @@ static void _transp_contribution_by_group(RealF use_by_group[]) {
                 LogError(logfp, LOGNOTE, "sxw_resource: Added transpiration less than 0.\n");
             }
-            //printf("Year %d:\tTranspiration to add: %f\n",Globals.currYear,add_transp);
+            //printf("Year %d:\tTranspiration to add: %f\n",Globals->currYear,add_transp);
```
This line can probably safely be removed.

chaukap replied:
@fdpierson I'm going to leave this print statement and remove the others, just so we have one for quick debugging.

fdpierson commented on sxw_resource.c:
```diff
@@ -300,101 +263,101 @@ static void _transp_contribution_by_group(RealF use_by_group[]) {
             }
-            //printf("Year %d:\tTranspiration to add: %f\n",Globals.currYear,add_transp);
+            //printf("Year %d:\tTranspiration to add: %f\n",Globals->currYear,add_transp);
             //printf("TranspRemaining: %f\tTranspRemaining+add_transp: %f\n",TranspRemaining,add_transp+TranspRemaining);
```
This line can probably safely be removed.

chaukap replied:
Resolved by 103ca00.

fdpierson commented on ST_seedDispersal.c:
```diff
@@ -0,0 +1,324 @@
+	printf("\nSeed dispersal during the simulation is not yet functional.\n"
+		    "Check out GitHub for updates on this feature.\n");
+	UseSeedDispersal = FALSE;
+	return;
```
Remove redundant return.

chaukap replied:
@fdpierson I'm going to leave this return in until we decide how to overhaul seed dispersal. I want to make it clear that this function does not work and should return immediately.

fdpierson commented on ST_seedDispersal.c:
```diff
@@ -0,0 +1,324 @@
+}
+
+void _do_precise_dispersal(int leftoverSeeds, SppIndex sp){
+    return;
```
Remove redundant return.

chaukap replied:
@fdpierson I added the explicit return to make it clear to future programmers that this function is intentionally empty until seed dispersal is implemented. I will, however, add a comment as well to make it clearer.

fdpierson commented on ST_stats.c:
```diff
@@ -26,97 +26,23 @@
 #include "sw_src/SW_Site.h"
 #include "ST_structs.h"
+#include "ST_stats.h" // Contains most of the function declarations.
+#include "ST_initialization.h"
```
Unused include. "stdlib.h" and "sw_src/SW_Site.h" are also unused.

chaukap replied:
Resolved by 5b1b919.

fdpierson commented on ST_stats.c:
```diff
@@ -26,97 +26,23 @@
+#include "ST_stats.h" // Contains most of the function declarations.
+#include "ST_initialization.h"
+#include "ST_seedDispersal.h"
 #include "sxw.h"
```
Unused include.

chaukap replied:
Resolved by 5b1b919.

fdpierson commented on ST_stats.c:
```diff
@@ -1480,18 +838,18 @@ static void _make_header_with_std( char *buf) {
 
 /***********************************************************/
-static void _make_header( char *buf) {
+void make_header( char *buf) {
```
Function is never used.

chaukap replied:
@fdpierson I think we should leave these functions in for now just in case, but I will open an issue so we remember to address unused functions in another branch.

chaukap replied:
I created issue #393 to document these functions.

fdpierson commented on ST_stats.h:
```diff
@@ -0,0 +1,50 @@
+                            StatType* newGsize, StatType* newGpr, StatType* newGmort, StatType* newGestab, 
+                            StatType* newSpp, StatType* newIndv, StatType* newSmort, StatType* newSestab, 
+                            StatType* newSrecieved, FireStatsType* newGwf, Bool firstTime);
+void make_header( char *buf);
```
Function is never used.

fdpierson commented on ST_stats.h:
```diff
@@ -0,0 +1,50 @@
+void stat_Collect( Int year ) ;
+void stat_Collect_GMort ( void ) ;
+void stat_Collect_SMort ( void ) ;
+void stat_Output_YrMorts( void ) ;
```
Function is never used. Address in ST_stats.c as well unless we want to keep it for some reason.

chaukap commented on ST_grid.c:
```diff
@@ -706,2357 +389,1198 @@ static void _init_grid_files(void)
 
-/***********************************************************/
+/* Read in the STEPWAT2 files and populate the grid. This only needs to be called once. 
+   DEPENDENCYS: gridCells must be allocated first. */
```
Resolved by 5b1b919.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,369 @@
 #include <string.h>
 #include <math.h>
 #include <ctype.h>
 #include <errno.h>
```
Resolved by 5b1b919.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,369 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <math.h>
```
Resolved by 5b1b919.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,369 @@
 /* --------------------------------------------------- */
-
 #include <stdio.h>
 #include <stdlib.h>
```
Resolved by 5b1b919.

chaukap commented on ST_grid.c:
```diff
@@ -706,2357 +389,1198 @@ static void _init_grid_files(void)
 static void _init_SXW_inputs(Bool init_SW, char *f_roots)
 {
 	SXW_Init(init_SW, f_roots);	// initializes soilwat
 	if (init_SW == TRUE)
```
Resolved by d3cb3fd.

chaukap commented on ST_indivs.c:
```diff
@@ -118,6 +119,30 @@ Bool indiv_New( SppIndex sp) {
   return( TRUE);
```
Resolved by d3cb3fd. I removed the if-statement and parenthesis.

chaukap commented on ST_main.c:
```diff
@@ -187,51 +190,44 @@ int main(int argc, char **argv) {
+                logProgress(iter, year, SIMULATION);
+            }
 
       //printf("------------------------Repetition/year = %d / %d\n", iter, year);
```
Resolved by d3cb3fd.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,369 @@
+				for (j = 0; j < grid_Cols; j++)
 				{ //for each cell
 
 					//fprintf(stderr, "year: %d", year);
```
Resolved by d3cb3fd.

chaukap commented on ST_main.c:
```diff
@@ -217,6 +217,8 @@ int main(int argc, char **argv) {
                 logProgress(iter, year, SIMULATION);
             }
 
+			//printf("------------------------Repetition/year = %d / %d\n", iter, year);
```
@kpalmqui Sorry about that, I added the statement back.

chaukap commented on ST_grid.c:
```diff
@@ -706,2357 +385,1198 @@ static void _init_grid_files(void)
-	if (UseSoils)
-		ChDir(grid_directories[0]); //change the directory for _init_soil_layers()
-	for (i = 0; i < nSoilTypes; i++)
+	//pretty much this function goes through s until it find nSeperators worth of seperators and then returns the index of the next character
```
Lets clean up this documentation and move it above the function.

chaukap commented on ST_grid.h:
```diff
@@ -0,0 +1,168 @@
+}typedef SoilType;
+
+/* Initialization information. */
+struct _grid_init_species_st
```
Remove the leading `_`. This struct is exported.

chaukap commented on ST_progressBar.c:
```diff
@@ -0,0 +1,115 @@
+/* Log the program's progress using a progress bar.
+	Param iteration: integer greater than 0. Input 0 iff the program is not currently in an iteration loop.
+	Param year: integer greater than 0. Input 0 iff the program is not currently in a years loop.
+	Param status: Use the "status" enum to choose a value.  */
```
Use the "Status" enum.

chaukap commented on ST_progressBar.c:
```diff
@@ -0,0 +1,115 @@
+void logProgress(int iteration, int year, Status status){
+	static char progressString[256];
+	int index = 0;					// Where we are in progressString
+	Bool needsProgressBar = FALSE;	// By default we donot need a progress bar
```
*we do not*

chaukap commented on ST_progressBar.c:
```diff
@@ -0,0 +1,115 @@
+     Param year: Current year in the loop.
+	 Param iteration: Current iteration in the loop.
+	 Param status: Use the "status" enumerator. Valid options are SPINUP or SIMULATION. */
+double _calculateProgress(int year, int iteration, Status status){
```
We should generalize the naming of this function. `year` and `iteration` are too specific. I'm thinking `innerLoopIteration` and `outerLoopIteration` to denote the loop structure.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,365 @@
+extern Bool* _SomeKillage;              // From ST_mortality.c
 
 /******** Modular External Function Declarations ***********/
 /* -- truly global functions are declared in functions.h --*/
```
Oops, I must have read the deleted lines on accident. This header is correct.

chaukap commented:
@kpalmqui I am now happy with the changes. Once you have a chance to review we can merge this branch and begin working on seed dispersal.

kpalmqui commented:
@chaukap I should be able to get this done by Friday.

kpalmqui commented on ST_grid.c:
```diff
@@ -706,2357 +390,1206 @@ static void _init_grid_files(void)
+    i = sscanf(buf, "%u", &UseDisturbances);
+    if (i != 1)
+        LogError(logfp, LOGFATAL,
+                 "Invalid grid setup file (distur
```
Let's ensure these comments have been translated over

kpalmqui commented on ST_grid.c:
```diff
@@ -706,2357 +390,1206 @@ static void _init_grid_files(void)
+    i = sscanf(buf, "%u", &UseDisturbances);
+    if (i != 1)
+        LogError(logfp, LOGFATAL,
+                 "Invalid grid setup file (distur
```
MAXDP and maxCells are not in the code anymore - we need to retain all of the old seed dispersal implementation so we can easily move forward in the next development branch. I am leaning towards retaining this whole function for the time being, so we don't lose any of the existing functionality, even though it is poorly implemented.

chaukap replied:
I added these variables back to the initDispersalParameters function and added the lines associated with them.
Resolved by 0291723.

kpalmqui commented on ST_grid.c:
```diff
@@ -706,2357 +390,1206 @@ static void _init_grid_files(void)
+    i = sscanf(buf, "%u", &UseDisturbances);
+    if (i != 1)
+        LogError(logfp, LOGFATAL,
+                 "Invalid grid setup file (distur
```
I think we should retain this function in the code, so we don't lose any existing information when we implement the overhaul to seed dispersal.

chaukap replied:
In ST_seedDispersal.c the function disperseSeeds() contains this entire function. It is commented out for now but it still exists so we can reference it.

kpalmqui commented on ST_grid.c:
```diff
@@ -706,2357 +390,1206 @@ static void _init_grid_files(void)
+    i = sscanf(buf, "%u", &UseDisturbances);
+    if (i != 1)
+        LogError(logfp, LOGFATAL,
+                 "Invalid grid setup file (distur
```
why has this been deleted?

chaukap replied:
The point of this function was to set lyppt to last year's precipitation before a new precipitation value is calculated. There were two reasons why I deleted this function.

1) It makes no sense to store precipitation for each species. Precipitation is a singe value so I only store it in one place. Therefore, iterating through species is no longer needed.
2) The grid_SD variable no longer exists. It is now stored inside the CellType struct so we only have to instantiate a singe 2d array. 

The whole function could be condensed into 
```
if (year > 1 && UseSeedDispersal){
    gridCells[i][j].mySeedDispersal->lyppt = gridCells[i][j].myEnvironment.ppt;
}
```
which I did and placed it in runGrid()

kpalmqui replied:
@chaukap thanks for the clarification!

chaukap commented on ST_indivs.c:
```diff
@@ -111,11 +112,32 @@ Bool indiv_New( SppIndex sp) {
 
-  //sql for inserting new indiv
-  //if(!UseGrid)
-	//  insertIndiv(p);
```
@kpalmqui That sounds good. I'll add back the sql lines and open an issue to revisit them.

chaukap replied:
Issue #391 has been opened to address this functionality. I also added the lines back in commit 715e1e3.

chaukap commented on ST_indivs.c:
```diff
@@ -111,11 +112,32 @@ Bool indiv_New( SppIndex sp) {
+
+/* Copy one individual's information to another individual. 
+   Note: this does not modify either individual's linked list functionality.
+   Both individuals MUST be allocated. */
```
@kpalmqui `copy_individual` does copy the values directly. However, it is only used to copy the spinup conditions to a new location for running the simulation. 

I needed a way to "freeze" the entire grid following spinup so it could be loaded multiple times, one for each iteration of the simulation. The easiest way was to do something like `GridCells ** spinupConditions = gridCells` then, when we want to load the spinup conditions we allocate gridCells again and call the copy functions like `copy_individual` to move the data over. 

The data must be copied to a new location in memory or the first simulation iteration would overwrite the spinup values and there would be no way to recover them for the second iteration.

kpalmqui replied:
@chaukap thanks

chaukap commented on ST_initialization.c:
```diff
@@ -0,0 +1,301 @@
+/***********************************************************************/
+// ST_initialization.c 
+//      Contains definitions of all functions related to initialization.
+//      The current initialization methods are _run_seed_initialization
```
@kpalmqui I believe that setting up the logic and structure for _run_seed_initialization will make it easier for future programmers to work with he code base. Since the function already exists and the logic behind when to call the function is in place a future programer would only have to work inside the _run_seed_initialization function as opposed to also figuring out when they should call the function and where they should put the function. 

One thing that I should definitely do is update the documentation regarding seed dispersal in ST_initialization.c to make it clear what a future programmer has to do.

Of course this is completely up to you. If you would like I could remove the function and logic entirely.

kpalmqui replied:
We have decided to retain the functionality/code for now and re-evaluate the decision in the next development branch.

chaukap commented on ST_initialization.c:
```diff
@@ -0,0 +1,301 @@
+//      it to the InitializationMethod enumerator in ST_initialization.h, 
+//      add your method to _read_grid_setup() of ST_grid.c, then add your 
+//      function to the switch statement in runInitialization(). To see
+//      an example initialization function check out _run_spinup().
```
@kpalmqui No problem!

chaukap commented on ST_initialization.c:
```diff
@@ -0,0 +1,301 @@
+    RandSeed(SuperGlobals.randseed, &grid_rng);
+    RandSeed(SuperGlobals.randseed, &markov_rng);
+
+    if (BmassFlags.yearly || MortFlags.yearly)
```
@kpalmqui this was a good catch! I tried running gridded mode with bmassflags.yearly ON and the program crashed. parm_Initialize is a non-gridded function, and it won't even work unless the program changes directories and cycles through all cells.

It looks to me like we don't need the call to parm_Initialize at all. When we start gridded mode we use gridded mode functions to read all of the input files before doing anything else. Do you know what the point of rereading parameters between iterations was? It seems like Plot_Initialize and SXW_reset are sufficient for resetting the program between iterations. I tried removing the call to parm_Initialize and the program completed with expected output. In addition, @fdpierson noted that calling parm_Initialize multiple times could result in memory leaks.

This is an issue that must be addressed before moving forward. If you have any information on why parm_Initialize is called multiple times let me know, otherwise I believe we can remove it from gridded mode.

chaukap replied:
I have opened issue #389 to further discuss this problem.

chaukap replied:
We have determined that the call to parm_Initialize can be removed entirely.
Resolved by commit d192856.

chaukap commented on ST_initialization.c:
```diff
@@ -0,0 +1,301 @@
+    rgroup_IncrAges(); 			// Increment ages of all plants
+    _kill_annuals(); 			// Kill annuals
+    _kill_maxage();             // Kill plants that reach max age
+    proportion_Recovery(); 		// Recover from any disturbances
```
@kpalmqui proportion_Recovery decides if the current year is a fire year before it recovers any biomass. Therefore we can call it every time and let the function itself determine if anything should be done.

The alternative would be to check it it was a fire year before calling the function, but I would advise against pushing code out of the function it is associated with.

chaukap replied:
This was my mistake, I was looking at ST_grid.c when I wrote the previous comment. We can remove this call to proportion_Recovery.

chaukap replied:
Resolved by commit d192856.

chaukap commented on ST_main.c:
```diff
@@ -33,13 +33,17 @@
 #include "sw_src/SW_Output_outtext.h"
 #include "sw_src/SW_Output_outarray.h"
 #include "sw_src/rands.h"
-extern SXW_t SXW;
```
SXW is already externed in ST_grid.h which is included in this file. externing it again in ST_main was redundant so I removed it.

kpalmqui replied:
@chaukap thanks!

chaukap commented on ST_progressBar.c:
```diff
@@ -0,0 +1,120 @@
+/*********************** Function Definitions *****************************/
+
+/* Log the program's progress using a progress bar.
+	Param iteration: integer greater than 0. Input 0 iff the program is not currently in an iteration loop.
```
@kpalmqui this actually isn't a typo: "iff" is short for "if and only if" which is [used in math](https://en.wikipedia.org/wiki/If_and_only_if) to mean "both are true or both are false". I can understand how this would be confusing to people from other sciences so I'll use the full expression. Sorry for the confusion!

kpalmqui replied:
@ chaukap ok thanks!

chaukap commented on ST_resgroups.c:
```diff
@@ -996,6 +993,88 @@ void rgroup_Extirpate(GrpIndex rg)
+void copy_rgroup(const GroupType* src, GroupType* dest){
+    int i;
+
+    // This would be very bad.
```
Of course sorry about that!

chaukap replied:
Addressed by commit 715e1e3

chaukap commented on ST_species.c:
```diff
@@ -252,6 +256,119 @@ SppIndex species_New(void)
+		destIndv->Next = NULL;
+	}
+}
+
```
Copying the IndivType itself is handled by copy_individual. However, the linked list of individuals is stored inside the SpeciesType struct. That is why I put the responsibility of modifying the linked list inside copy_species.

chaukap commented on ST_stats.h:
```diff
@@ -0,0 +1,50 @@
+void stat_Output_Seed_Dispersal(const char * filename, const char sep, Bool makeHeader);
+void stat_free_mem( void );
+void stat_Copy_Accumulators(StatType* newDist, StatType* newPpt, StatType* newTemp, StatType* newGrp, 
+                            StatType* newGsize, StatType* newGpr, StatType* newGmort, StatType* newGestab, 
```
Unfortunately this isn't a typo. It's just a (poor) naming choice that was made in ST_stats.c. Grp is the group biomass and Gpr is the group PR. 

Personally I think programmers should never use acronyms and abbreviations in variable names because of confusing thinks like this. Even though changing `Grp` to `groupBiomass` makes the name longer most IDEs will recognize what you are trying to type within three characters and automatically fill it in.

But, I think the naming convention issue would be better addressed in another branch. For now I understand what they mean and the function works.

"There are only two hard things in Computer Science: cache invalidation and naming things." - Phil Karlton

kpalmqui replied:
@chaukap thanks for the clarification! I agree. And I appreciate the quote as well. :)

chaukap commented on ST_environs.c:
```diff
@@ -60,12 +60,55 @@ void Env_Generate( void) {
 /* mean, stddev that is between min & max from */
-/* the Globals.ppt structure.*/
+/* the Globals->ppt structure.*/
 /* Also set the growing season precip. */
```
Addressed by commit 715e1e3

chaukap commented on ST_environs.c:
```diff
@@ -74,7 +117,7 @@ static void _make_ppt( void) {
+ * KAP 1/26/2017 The above note by CB is not correct. Env->ppt
  *      is set in _sxw_set_environs, but gsppt is not, it is
  *      set below. When using SOILWAT or not, the gsspt, ppt.dry,
  *      and ppt.wet is currently fixed each year and read from env.in
```
Addressed by commit 715e1e3

chaukap commented on ST_grid.h:
```diff
@@ -0,0 +1,168 @@
+    enumerators from ST_grid.c. If you are looking for the functions 
+    that this module exports check the bottom of the file.
+
+    Initial programming by Chandler Haukap. */
```
Addressed by commit 715e1e3

chaukap commented on ST_grid.h:
```diff
@@ -0,0 +1,168 @@
+
+/*********************** Grid Structures ****************************/
+
+//represents the input data for all the soil layers of a cell
```
Addressed by commit 715e1e3

chaukap commented on ST_grid.h:
```diff
@@ -0,0 +1,168 @@
+/* Holds all plot-specific parameters */
+struct grid_cell_st
+{
+	/* RGroup coresponding to this cell */
```
Addressed by commit 715e1e3

chaukap commented on ST_grid.h:
```diff
@@ -0,0 +1,168 @@
+char *grid_files[N_GRID_FILES];
+/* Array of directory names. Use the Directory_Indices enum to pick the correct index. */
+char *grid_directories[N_GRID_DIRECTORIES];
+/* TRUE if every cell should write its own summary file. */
```
Addressed by commit 715e1e3

chaukap commented on ST_initialization.c:
```diff
@@ -0,0 +1,301 @@
+
+/***************************** Externed variables **********************************/
+/* Note that in an ideal world we wouldn't need to extern any variables because 
+   every module would declate them in a header file. Hopefully we can get this
```
Addressed by commit 715e1e3

chaukap commented on ST_initialization.c:
```diff
@@ -0,0 +1,301 @@
+
+}
+
+/* "Spinup" the model by running without stat collection. */
```
Addressed by commit 715e1e3

chaukap commented on ST_initialization.h:
```diff
@@ -0,0 +1,41 @@
+    Initialization is currently only availible when running gridded 
+    mode. 
+    
+    Initial programming by Chandler Haukap. */
```
Addressed by commit 715e1e3

chaukap commented on ST_params.c:
```diff
@@ -242,47 +224,48 @@ static void _env_init( void) {
             nitems = 8;
             break;
         case 2:
+            nitems = 5;
```
@kpalmqui you are correct. I removed one of the assignments.
Addressed by commit 715e1e3

chaukap commented on ST_params.c:
```diff
@@ -1207,39 +1190,45 @@ static void _species_init( void) {
+    Species[sp]->sd_Pmax = p5;
+    Species[sp]->sd_H = p6;
+    Species[sp]->sd_VT = p7;
+    Species[sp]->sd_VW = VW;
```
Addressed by commit 715e1e3 (variable now called p8)

chaukap commented on ST_progressBar.c:
```diff
@@ -0,0 +1,120 @@
+
+/* Log the program's progress using a progress bar.
+	Param iteration: integer greater than 0. Input 0 iff the program is not currently in an iteration loop.
+	Param year: integer greater than 0. Input 0 iff the program is not currently in a years loop.
```
Addressed by commit 715e1e3

chaukap commented on ST_progressBar.c:
```diff
@@ -0,0 +1,120 @@
+    Function definitions for a progress bar printed to the terminal.
+    See ST_progressBar.h for a description of how to add a new Status.
+ */
+/**************************************************************************/
```
Addressed by commit 715e1e3

chaukap commented on ST_progressBar.h:
```diff
@@ -0,0 +1,34 @@
+        progress bar, create a function for calculating progress
+        in ST_progressBar.c then add your function to 
+        _calculateProgress().
+*/
```
Addressed by commit 715e1e3

chaukap commented on ST_progressBar.h:
```diff
@@ -0,0 +1,34 @@
+
+/*********************** Enumerator(s) ****************************/
+
+/* States of the program that the progress bar recognises. */
```
Addressed by commit 715e1e3

chaukap commented on ST_seedDispersal.c:
```diff
@@ -0,0 +1,325 @@
+   and s is the species. */
+void initDispersalParameters(void)
+{
+	int senderRow, senderCol, recieverRow, recieverCol;
```
Addressed by commit 715e1e3

chaukap commented on ST_seedDispersal.c:
```diff
@@ -0,0 +1,325 @@
+   results of this function can be accessed using 
+   gridCells[a][b].mySeedDispersal[s].dispersalProb[c][d] 
+   where (a,b) are the coordinates of the sender,
+   (b, c) are the coordinates of the reciever,
```
Addressed by commit 715e1e3

chaukap commented on ST_seedDispersal.c:
```diff
@@ -0,0 +1,325 @@
+	double MAXD; /* maximum seed dispersal distance for a given species */
+	double maxRate; /* Dispersability of the seeds */
+	double plotWidth; /* width of the plots (all plots are equal and square) */
+	double distanceBetweenPlots; /* distance between the sender and the reciever */
```
Addressed by commit 715e1e3

chaukap commented on ST_seedDispersal.c:
```diff
@@ -0,0 +1,325 @@
+	/* sender denotes that these loops refer to the cell distributing seeds */
+	for(senderRow = 0; senderCol < grid_Rows; ++senderRow){
+		for(senderCol = 0; senderCol < grid_Cols; ++senderCol){
+			/* Cell is loaded to ensure the ForEachSpecies loop works */
```
Addressed by commit 715e1e3

chaukap commented on ST_seedDispersal.c:
```diff
@@ -0,0 +1,325 @@
+				}
+
+				/* These are the three values we need to calculate the probability of dispersal */
+				MAXD = ((Species[sp]->sd_H * Species[sp]->sd_VW) / Species[sp]->sd_VT) / 100.0; // divided by 100 to convert from cm to m.
```
Addressed by commit 715e1e3

chaukap commented on ST_seedDispersal.c:
```diff
@@ -0,0 +1,325 @@
+
+                        /* The value that we are after should be saved to the sender cell. */
+						sender->mySeedDispersal[sp].dispersalProb[recieverRow][recieverCol]
+                                    = (distanceBetweenPlots > MAXD) ? (0.0) : (exp(-maxRate * distanceBetweenPlots));
```
Addressed by commit 715e1e3

chaukap commented on ST_seedDispersal.c:
```diff
@@ -0,0 +1,325 @@
+	unload_cell();
+}
+
+/* Perform seed dispersal durring the simulation. This is NOT functional yet. */
```
Addressed by commit 715e1e3

chaukap commented on ST_stats.h:
```diff
@@ -0,0 +1,50 @@
+ * without including ST_stats.c. 
+ * 
+ * To read the documentation related to creating this struct see issues
+ * #259 and #266 on GitHub */
```
Addressed by commit 715e1e3

chaukap commented on ST_structs.h:
```diff
@@ -123,13 +123,14 @@ struct species_st {
+        sd_Pmax,      /* for seed dispersal */
+        sd_H,         /* for seed dispersal */
+        sd_VT,        /* for seed dispersal */
+        sd_VW;        /* for seed dispersal */
```
Issue #390 created to address this issue.

chaukap commented on ST_initialization.c:
```diff
@@ -0,0 +1,301 @@
+    unload_cell(); // Reset the global variables
+
+    for (year = 1; year <= SuperGlobals.runInitializationYears; year++)
+    { //for each year
```
Comment added in commit 715e1e3

chaukap commented on ST_initialization.h:
```diff
@@ -0,0 +1,41 @@
+    This header defines the functions exported from ST_initialization.c. 
+    Initialization is currently only availible when running gridded 
+    mode. 
+    
```
@kpalmqui I just realized we forgot to discuss this in our last call. I opened issue #392 so we can continue the discussion there.

chaukap commented on ST_species.c:
```diff
@@ -252,6 +256,119 @@ SppIndex species_New(void)
+	}
+
+	/* ---- Reallocate and copy any arrays ---- */
+	// kills: Note that this array is allocated iff MortFlags.summary.
```
Resolved by commit d192856.

chaukap commented on sxw.h:
```diff
@@ -15,6 +15,10 @@
 
+#define SXW_NFILES 5
+// The number of transpiration values retained by transp_data
+#define MAX_WINDOW 100
```
transp_window is already in env.in. MAX_WINDOW just caps the value a user can input to 100 to prevent excessive memory usage. See issue #169.

chaukap commented on sxw_resource.c:
```diff
@@ -237,29 +203,25 @@ static void _transp_contribution_by_group(RealF use_by_group[]) {
-      if(Globals.transp_window > MAX_WINDOW){
-                LogError(logfp, LOGNOTE, "sxw_resource: Transp_window specified in inputs is greater than maximum window.\nInput: %d\nMaximum: %d\nSetting window to %d.\n",Globals.transp_window, MAX_WINDOW,MAX_WINDOW);
-                Globals.transp_window = MAX_WINDOW;
-      }
```
@kpalmqui the LogError has been moved to _allocate_memory in sxw.c(). It made more sense to place it in _allocate_memory because it is only called once, whereas _transp_contribution_by_group is called multiple times. transp_window only has to be checked and set once.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,370 @@
+        // i.e. fills the accumulators in ST_stats.c.
+		if (MortFlags.summary){
+			for (i = 0; i < grid_Rows; i++){
+				for (j = 0; j < grid_Cols; j++)
```
@kpalmqui correct. I changed the 1-indexing to 0-indexing because all C arrays are 0-indexed, so a C programer should expect to begin iterations at 0.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,370 @@
-			for (year = 2; year <= Globals.runModelYears; year++) // _load_cell gets the first years accumulators loaded, so we start at 2...
-				stat_Load_Accumulators(cell, year);
-
+			int cell = (j + 1) + (i * grid_Cols) - 1;
```
@kpalmqui deleted line 551 and added line 321 do the exact same thing, except 551 expects 1-indexed loops and 321 expects 0-indexed loops. 

However, There is an inefficiency in the new code:
```
(j + 1) + (i * grid_Cols) - 1
= (j + 1) - 1 + (i * grid_Cols)
=  j + 1 - 1 + (i * grid_Cols)
= j + (i * grid_Cols)
```
I will definitely fix this.

With the new reduced code it is clear that the two lines are equivalent:
```
Original:
j + ((i - 1) * grid_Cols) - 1
= (j - 1) + ((i - 1) * grid_Cols)

New: 
j + (i * grid_Cols)
```

So the only difference is the old implementation is subtracting 1 from i and j to account for 1-indexing.

chaukap replied:
I simplified the line to `j + (i * grid_Cols)`
Resolved by 0291723.

chaukap commented on ST_grid.c:
```diff
@@ -706,2357 +390,1206 @@ static void _init_grid_files(void)
+                      										   "_allocate_accumulators(Gwf->prescribedFire)");
+      				}
+    			}
+  			}
```
@kpalmqui this flag is correctly placed. BmassFlags.grpb determines whether or not group statistics should be printed. Since prescribed fire is printed by group it only needs to be allocated if BmassFlags.grpb is on. 

MortFlags.group is roughly the equivalent flag for the mortality output file.

chaukap replied:
The reason the spacing is strange is a mixing of spaces and tabs, which should be fixed when running clang formatter on the project.

kpalmqui replied:
@chaukap ok thanks!

chaukap commented on ST_grid.c:
```diff
@@ -706,2357 +390,1206 @@ static void _init_grid_files(void)
+                          						"_allocate_accumulators(Indv[sp].s)");
+					}
+    			}
+  			}
```
The bracket structure is correct, but irregularities in indentation make it look strange. Someone (possibly me) used 2 spaces followed by a tab before the bracket in question which is displayed differently by every document viewer. Again, this will be fixed by running clang formatter.

chaukap commented on ST_main.c:
```diff
@@ -60,20 +64,14 @@ SW_FILE_STATUS SW_File_Status;
-
-  void stat_Output_AllMorts( void) ;
-  void stat_Output_AllBmass(void) ;
-
```
@kpalmqui All of these functions are now declared in ST_stats.h so there is no need to redeclare them here.

chaukap commented on ST_stats.c:
```diff
@@ -476,251 +322,28 @@ static void _init( void) {
 
-/***********************************************************/
-void stat_Init_Accumulators( void ) {
-	//allocates memory for all of the grid accumulators
```
Resolved by 0291723.

chaukap commented on ST_stats.c:
```diff
@@ -476,251 +322,28 @@ static void _init( void) {
-}
-
-/***********************************************************/
-void stat_Load_Accumulators(int cell, int year) {
```
Resolved by 0291723.

chaukap commented on ST_stats.c:
```diff
@@ -476,251 +322,28 @@ static void _init( void) {
-}
-
-/***********************************************************/
-void stat_Save_Accumulators(int cell, int year) {
```
Resolved by 0291723.

chaukap commented on ST_stats.c:
```diff
@@ -476,251 +322,28 @@ static void _init( void) {
-}
-
-/***********************************************************/
-void stat_Free_Accumulators( void ) {
```
Resolved by 0291723.

chaukap commented on ST_grid.c:
```diff
@@ -706,2357 +390,1206 @@ static void _init_grid_files(void)
+{
+	//this is similar to the getaline function in filefuncs.c, except this one checks for carriage return characters and doesn't deal with whitespace/... (since excel writes them into .csv files for some aggravating reason)... this one is probably less efficient overall though.
+	//this is treating '\r', '\n', and '\r\n' all like they are valid line feed characters... in reality '\r' by itself should never be, but it's not like I can stop excel from outputting .csv files however the heck it feels like...
+	//only read limit characters
```
Resolved by 0291723.

chaukap commented on sxw.c:
```diff
@@ -1124,68 +1150,47 @@ int getNTranspLayers(int veg_prod_type) {
-}
-
-/***********************************************************/
-void free_sxw_memory( void ) {
```
Resolved by 0291723.

chaukap commented on sxw.h:
```diff
@@ -96,48 +97,77 @@ struct transp_data {
+  /* simple vectors hold the resource information for each group */
+  /* curr/equ gives the available/required ratio */
+  RealF *_resource_cur,  /* current resource availability for each STEPPE functional type */
+        *_resource_pr;   /* resource convertable to PR */
```
Resolved by 0291723.

chaukap commented on sxw_resource.c:
```diff
@@ -47,46 +47,15 @@
-
-extern
-  RealF *_resource_cur; /* resources currently available by group*/
-        RealF added_transp; /* transpiration added for the current year */
```
Resolved by 0291723.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,370 @@
+// Source file: ST_grid.c
+// Type: module
+// Application: STEPPE - plant community dynamics simulator
+// Purpose: This module handles the grid.
```
I changed the comment to "This module performs gridded mode simulations."
Resolved by 0291723.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,370 @@
-						_set_sd_lyppt(i, j);
+					rgroup_Establish(); 		// Establish individuals. Excludes annuals.
 
+					Env_Generate();				// Generated the SOILWAT environment
```
Resolved by 0291723.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,370 @@
 
-					if (year > 1 && UseSeedDispersal)
-						_set_sd_lyppt(i, j);
+					rgroup_Establish(); 		// Establish individuals. Excludes annuals.
```
Resolved by 0291723.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,370 @@
 					}
 
-					Globals.currYear = year;
+					/* The following functions mimic ST_main.c. load_cell(i, j) ensures that all global variables reference the specific cell */
```
Resolved by 0291723.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,370 @@
+					Env_Generate();				// Generated the SOILWAT environment
 
+					rgroup_PartResources();		// Distribute resources
+					rgroup_Grow(); 				// Grow
```
Resolved by 0291723.

chaukap commented on ST_grid.c:
```diff
@@ -706,2357 +390,1206 @@ static void _init_grid_files(void)
+
+			// Set mySXW to the location of the newly allocated SXW
+			gridCells[i][j].mySXW = getSXW();	
+			// Set myTanspWindow to the location of the newly allocated transp window
```
Resolved by 0291723.

chaukap commented on sxw.c:
```diff
@@ -82,42 +82,21 @@ extern SW_MARKOV SW_Markov;
+SXW_resourceType* SXWResources;
+
+#ifdef SXW_BYMAXSIZE
+/* addition to meet changes specified at the top of the file */
```
I investigated the variable defined in this block, `_Grp_BMass`, and I can't find it anywhere in the code. The only other reference in the code is at the top of sxw.c explaining why it exists. Apparently this feature is deprecated, so I'll remove it entirely.

One side note: While this block appears to be new code it is not. The only reason it is showing up as "new" is because I moved a lot of code that was above it. Git is fairly good at recognizing things like this, but in this case it failed to associate new line 97 with it's previous location.

chaukap commented on testing.sagebrush.master/Grid_Inputs/grid_setup.in:
```diff
@@ -3,4 +3,12 @@
+
+#SEED DISPERSAL OUTPUTS
+1		# output seed dispersal summary file?  1 means yes, 0 means no
+1		# output header?  1 means yes, 0 means no.
```
I'll remove this parameter and adjust the code to create a header automatically.

chaukap replied:
Addressed by commit 0291723.

chaukap commented on ST_structs.h:
```diff
@@ -284,21 +285,12 @@ struct globals_st {
         tempparm[2][3]; /* three parms for Warm/Cool growth mod*/
-  IntUS runModelYears,  /* number of years to run the model*/
-      Max_Age,        /* oldest plant; same as runModelYears for now */
+  IntUS Max_Age,        /* oldest plant; same as runModelYears for now */
```
@kpalmqui It is used when outputting mortality information. I'm not sure why it is used as opposed to Globals->runModelYears since they are the same, but my impression is that someone was going to set Max_Age to species-specific values but never got around to it.

kpalmqui replied:
@chaukap ok -thanks!

chaukap commented on ST_main.c:
```diff
@@ -440,7 +497,6 @@ static void init_args(int argc, char **argv) {
   /* Defaults */
   parm_SetFirstName( DFLT_FIRSTFILE);
   QuietMode = EchoInits = UseSeedDispersal = FALSE;
-  SXW.debugfile = NULL;
```
Since I changed SXW to a pointer it is not allocated when this function runs. This line has been moved to SXW_Init right after the memory is allocated.

kpalmqui commented on ST_environs.c:
```diff
@@ -60,27 +60,67 @@ void Env_Generate( void) {
+ * KAP 1/26/2017 We should consider calculating amount of growing-season 
+ *      precipitation this year in the _sxw_set_environs function when 
+ *      running SOILWAT (so it is not fixed), and allowing what constitutes 
+ *      a wet and dry year to vary across sites. */
```
this comment needs to be deleted as gsppt is set down on the fly in _make_ppt

chaukap replied:
Addressed by commit 6a53657.

kpalmqui commented on ST_grid.c:
```diff
@@ -1,686 +1,370 @@
-struct _grid_sd_struct
-{ //for seed dispersal
-	//the idea is that this structure contains all of the cells that the cell represented can possibly disperse seeds to and the probability of the happening for each cell
-	int size, seeds_present, seeds_received, *cells; //seeds_present && seeds_received are treated as boolean values... cells & prob are to be the length of size
```
@chaukap thanks!

chaukap commented on sxw_sql.c:
```diff
@@ -339,9 +337,9 @@ void insertOutputVars(RealF * _resource_cur, RealF added_transp) {
-		//ForEachTranspLayer(t) sum += SXW.transp[Ilp(t,p)];
-		for (t = 0; t < SXW.NSoLyrs; t++)
-			sum += SXW.transpTotal[Ilp(t, p)];
+		//ForEachTranspLayer(t) sum += SXW->transp[Ilp(t,p)];
```
Addressed by commit 0291723.

chaukap commented on ST_grid.c:
```diff
@@ -706,2357 +390,1206 @@ static void _init_grid_files(void)
+                         							sizeof(struct accumulators_st),
+                        							"_allocate_accumulators(GSize[rg].s)");
+					}
+    			}
```
Addressed by commit 0291723.

chaukap commented on testing.sagebrush.master/Grid_Inputs/grid_initSpecies.csv:
```diff
@@ -1 +1 @@
140,1,4,,,,,,,,,,,,,,,,,,,
141,1,4,,,,,,,,,,,,,,,,,,,
142,1,4,,,,,,,,,,,,,,,,,,,
143,1,4,,,,,,,,,,,,,,,,,,,
```
You're right; I'll reduce the number of cells to 16 so a user can run a 4x4 grid before they have to modify grid_initSpecies.csv

chaukap replied:
Resolved by commit f70d4af.

chaukap commented on testing.sagebrush.master/Stepwat_Inputs/Input/species.in:
```diff
@@ -173,29 +173,30 @@
 # Pmax = probability of producing seeds in a wet year
 # H = the average release height of the inflorescences (cm) 
 # VT = the average sinking velocity of the seeds (cm/sec)
+# VW - the mean horizontal wind speed (500 cm/sec in the paper)
```
Addressed by commit 6a53657.

chaukap commented on testing.sagebrush.master/Grid_Inputs/grid_soils.csv:
```diff
@@ -1 +1,41 @@
7,1,1,0,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
8,1,1,0,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
\ No newline at end of file
+cell,copy_cell,num_layers,layer,depth,matricd,gravel_content,evco,trco_grass,trco_shrub,trco_tree,trco_forb,sand,clay,imperm,soiltemp,sxw_roots
```
Resolved by commit f70d4af.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,370 @@
+		   file names. */
+		ChDir(grid_directories[GRID_DIRECTORY_STEPWAT_INPUTS]);
+		SXW_Reset(gridCells[0][0].mySXW->f_watin);
+		ChDir("..");
```
@kpalmqui I expanded the comment to explain the code in the best detail I can. It appress to be caused by the fact that we run gridded mode from the testing.sagebrush.master directory but SXW expects to be run from the testing.sagebrush.master/Stepwat_Inputs directory.

See commit 01db611.

chaukap commented on ST_seedDispersal.c:
```diff
@@ -0,0 +1,325 @@
+			sender->mySeedDispersal = Mem_Calloc(MAX_SPECIES, sizeof(Grid_SD_St), "initDispersalParameters");
+
+			ForEachSpecies(sp){
+				if (!(Species[sp]->use_me && Species[sp]->use_dispersal)){
```
We need the second statement because some species could be on for the simulation but off for seed dispersal. If we removed the second condition it would be impossible to turn off seed dispersal for a specific species.

kpalmqui replied:
@chaukap thanks for the clarification!

chaukap commented:
@kpalmqui I believe I have addresses your reviews entirely. If you want any more clarification on any of my comments please let me know or we can discuss them on our next call. 

The only remaining issue is seed dispersal. Have you decided whether or not we should retain the seed dispersal code for now?

kpalmqui commented:
@chaukap great thank you! I will re-review. Let's retain seed dispersal and deal with it in the next branch as originally planned.

kpalmqui commented on ST_initialization.c:
```diff
@@ -0,0 +1,301 @@
+    initializationMethod = INIT_WITH_NOTHING;
+	return;
+}
+
```
OK, let's retain for now and revisit in the next development branch focused on seed dispersal.

kpalmqui commented on testing.sagebrush.master/Grid_Inputs/grid_setup.in:
```diff
@@ -3,4 +3,12 @@
 1		# use soils csv file (0 or 1)... 0 means no, 1 means yes
-1		# use seed dispersal (0 or 1)... 0 means no, 1 means yes
+0		# use seed dispersal (0 or 1)... 0 means no, 1 means yes
+spinup  	# Initialization method. Options are "spinup", "seeds", or "none"
```
keep for now. resolve in seed dispersal development branch.

chaukap commented on ST_grid.c:
```diff
@@ -1,686 +1,367 @@
 void rgroup_Establish(void);
 void rgroup_IncrAges(void);
 void rgroup_PartResources(void);
-//void rgroup_ResPartIndiv(void);
```
@kpalmqui That is correct. We don't need to call this function directly from ST_grid.c so there is no reason to declare it here.

kpalmqui replied:
@chaukap thanks!

chaukap commented:
@kpalmqui I just ran 6 single-plot simulations on both master and gridded code. Heres the parameters I used:
1) 1 iteration 300 years
2) 3 iterations 300 years
3) 3 iterations 300 years with wildfire
4) 3 iterations 300 years with stochastic prescribed fire for all rgroups.
5) 1 iteration 300 years with prescribed fire for sagebrush only.
6) 1 iteration 300 years with grazing turned on for a.cool.forb

To compare the output I looked at the average values for temperature, precipitation, sagebrush biomass and a.cool.forb biomass. In all cases the values were Identical. To double check, in the simulations with fire I checked the years that fire occurred and they were also identical. 

I therefore conclude that single-plot mode has not been altered by the changes on this branch.

chaukap commented:
@kpalmqui Thank you for all your guidance! I'm waiting for the Appveyor checks to clear then I'll get this branch folded in. The companion SOILWAT2 and rSFSTEP2 branches have been merged in already.
